src/libstore/worker-protocol.hh

```diff
@@ -66,14 +66,104 @@ typedef enum {
 class Store;
 struct Source;
 
-template<class T> T readStorePaths(const Store & store, Source & from);
-
-void writeStorePaths(const Store & store, Sink & out, const StorePathSet & paths);
-
-StorePathCAMap readStorePathCAMap(const Store & store, Source & from);
-
-void writeStorePathCAMap(const Store & store, Sink & out, const StorePathCAMap & paths);
-
-void writeOutputPathMap(const Store & store, Sink & out, const OutputPathMap & paths);
+template<typename T>
+struct WorkerProto {
+    static T read(const Store & store, Source & from);
+    static void write(const Store & store, Sink & out, const T & t);
+};
+
+#define MAKE_WORKER_PROTO(T) \
+    template<> \
+    struct WorkerProto< T > { \
+        static T read(const Store & store, Source & from); \
+        static void write(const Store & store, Sink & out, const T & t); \
+    }
+
+MAKE_WORKER_PROTO(std::string);
+MAKE_WORKER_PROTO(StorePath);
+MAKE_WORKER_PROTO(ContentAddress);
+
+template<typename T>
+struct WorkerProto<std::set<T>> {
+
+    static std::set<T> read(const Store & store, Source & from)
+    {
+        std::set<T> resSet;
+        auto size = readNum<size_t>(from);
+        while (size--) {
+            resSet.insert(WorkerProto<T>::read(store, from));
+        }
+        return resSet;
+    }
+
+    static void write(const Store & store, Sink & out, const std::set<T> & resSet)
+    {
+        out << resSet.size();
+        for (auto & key : resSet) {
+            WorkerProto<T>::write(store, out, key);
+        }
+    }
+
+};
+
+template<typename K, typename V>
+struct WorkerProto<std::map<K, V>> {
+
+    static std::map<K, V> read(const Store & store, Source & from)
+    {
+        std::map<K, V> resMap;
+        auto size = readNum<size_t>(from);
+        while (size--) {
+            auto k = WorkerProto<K>::read(store, from);
+            auto v = WorkerProto<V>::read(store, from);
+            resMap.insert_or_assign(std::move(k), std::move(v));
+        }
+        return resMap;
+    }
+
+    static void write(const Store & store, Sink & out, const std::map<K, V> & resMap)
+    {
+        out << resMap.size();
```

Comment:
If I'm not mistaken, this breaks an implicit assumption in the protocol that the first line of every non-scalar “packet” indicates the size (in number of lines) of the payload.

I'm not sure how important this is in practice − it's conceptually nicer as it means that the outer syntax of the protocol isn't context dependent and might prevent some ugly failures when things go bad − but I think it's worth mentioning (ftr the choice of using `2*map.size()` in the first iteration was guided by that)

Reply:
Ah OK. I didn't realize your intent there. I'm fine either way.

Reply:
Either am I tbh. I just wanted to point that out in case that was a deliberate design choice by @edolstra

Reply:
I think this ended up being done in an earlier PR?